HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlgo.dev.cc

```diff
@@ -8,35 +8,45 @@
 #include "DataFormats/PortableTestObjects/interface/alpaka/TestDeviceCollection.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/config.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/traits.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
 
 #include "TestAlgo.h"
 
 namespace ALPAKA_ACCELERATOR_NAMESPACE {
 
+  using namespace cms::alpakatools;
+
   class TestAlgoKernel {
   public:
-    template <typename TAcc, typename = std::enable_if_t<cms::alpakatools::is_accelerator_v<TAcc>>>
+    template <typename TAcc, typename = std::enable_if_t<is_accelerator_v<TAcc>>>
     ALPAKA_FN_ACC void operator()(TAcc const& acc, portabletest::TestDeviceCollection::View view, int32_t size) const {
-      // this example accepts an arbitrary number of blocks and threads, and always uses 1 element per thread
+      // global index of the thread within the grid
       const int32_t thread = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0u];
-      const int32_t stride = alpaka::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc)[0u];
+      const portabletest::Matrix matrix{{1, 2, 3, 4, 5, 6}, {2, 4, 6, 8, 10, 12}, {3, 6, 9, 12, 15, 18}};
+
+      // set this only once in the whole kernel grid
       if (thread == 0) {
         view.r() = 1.;
       }
-      for (auto i = thread; i < size; i += stride) {
-        view[i] = {0., 0., 0., i};
+
+      // make a strided loop over the kernel grid, covering up to "size" elements
+      for (int32_t i : elements_with_stride(acc, size)) {
```

Comment:
Just a thought for future, not customizing the for-loop itself makes it less clean to add `#pragma ivdep` or `#pragma omp simd` for the CPU backends (e.g. Kokkos uses `#pragma ivdep` and requires the `parallel_for` loop elements to be independent). Correctness of both of course depend on what happens inside the loop, so one option would be to define our own macro (yay) for a user to declare a loop to have independent iterations.

Reply:
Mhm.. I see.

We do have `FWCore/Utilities/interface/CMSUnrollLoop.h`

Reply:
By the way, from a quick poll during the hackathon people seem to prefer the explicit for loop approach, and I think it does give a bit more flexibility.

However, we do have loops that take the body as a lambda (or other function object) in pixeltrack-standalone, so we should re-check them and import them in CMSSW as well.